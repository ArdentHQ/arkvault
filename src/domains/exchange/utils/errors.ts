interface ErrorWithResponse extends Error {
	response: any;
}

const isInvalidAddressError = (error: ErrorWithResponse) => {
	try {
		return error.response.json().error.message === "Invalid Address";
	} catch {
		return false;
	}
};

const isInvalidRefundAddressError = (error: ErrorWithResponse) => {
	try {
		return error.response.json().error.message === "Invalid Refund Address";
	} catch {
		return false;
	}
};

const isUnavailablePairError = (error: ErrorWithResponse) => {
	try {
		return error.response.json().error.message === "Unavailable Pair";
	} catch {
		return false;
	}
};

export { isInvalidAddressError, isInvalidRefundAddressError, isUnavailablePairError };
