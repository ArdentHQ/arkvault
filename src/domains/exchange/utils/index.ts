export * from "./assertions";
export * from "./errors";
