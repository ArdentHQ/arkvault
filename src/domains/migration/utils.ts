import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
