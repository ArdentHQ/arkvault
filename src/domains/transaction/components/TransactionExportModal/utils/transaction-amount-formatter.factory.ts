import { BigNumber } from "@ardenthq/sdk-helpers";
import { DTO, Helpers } from "@ardenthq/sdk-profiles";

const formatAmount = (amount: number, transaction: DTO.ExtendedConfirmedTransactionData) =>
	Helpers.Currency.format(amount, transaction.wallet().exchangeCurrency(), {
		withTicker: false,
	});

const multiPaymentAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isReceived()) {
		let totalReceived = BigNumber.make(transaction.amount());

		for (const recipient of transaction.recipients()) {
			if (recipient.address === transaction.wallet().address()) {
				totalReceived.minus(recipient.amount);
			}
		}

		return formatAmount(totalReceived.toNumber(), transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return multiPaymentAmount(transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionTotal = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return BigNumber.make(multiPaymentAmount(transaction)).plus(transaction.fee());
	}

	return formatAmount(transaction.total(), transaction);
};

export const AmountFormatter = (transaction: DTO.ExtendedConfirmedTransactionData) => ({
	amount: () => transactionAmount(transaction),
	fee: () => formatAmount(transaction.fee(), transaction),
	total: () => transactionTotal(transaction),
});
