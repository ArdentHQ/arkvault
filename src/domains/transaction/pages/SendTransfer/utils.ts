export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
