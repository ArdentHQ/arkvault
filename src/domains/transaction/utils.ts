import { Services } from "@ardenthq/sdk";
import { RecipientItem } from "@/domains/transaction/components/RecipientList/RecipientList.contracts";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const getTransferType = ({ recipients }: { recipients: RecipientItem[] }): "multiPayment" | "transfer" => {
	const isMultiPayment = recipients.length > 1;
	return isMultiPayment ? "multiPayment" : "transfer";
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
