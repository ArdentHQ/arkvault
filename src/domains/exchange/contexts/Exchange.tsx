import { camelCase } from "@payvo/sdk-helpers";
import React, { useCallback, useMemo, useState } from "react";

import { httpClient } from "@/app/services";
import { Exchange } from "@/domains/exchange/exchange.contracts";
import { exchangeHost, ExchangeService } from "@/domains/exchange/services/exchange.service";

interface Properties {
	children: React.ReactNode;
}

const ExchangeContext = React.createContext<any>(undefined);

const useExchange = () => {
	const [provider, setProvider] = useState<Exchange | undefined>();
	const [exchangeProviders, setExchangeProviders] = useState<Exchange[] | undefined>();

	const exchangeService = useMemo<ExchangeService | undefined>(() => {
		if (provider) {
			return new ExchangeService(provider.slug, httpClient);
		}
	}, [provider]);

	const fetchProviders = useCallback(async () => {
		const transformExchanges = (exchanges: Exchange[]) =>
			exchanges.map((exchange: Exchange) => {
				const newExchange: any = {};

				for (const [key, value] of Object.entries(exchange)) {
					newExchange[camelCase(key) as string] = value;
				}

				return newExchange;
			});

		try {
			const body = await httpClient.get(exchangeHost);
			setExchangeProviders(transformExchanges(body.json().data));
		} catch {
			setExchangeProviders([]);
		}
	}, []);

	return {
		exchangeProviders,
		exchangeService,
		fetchProviders,
		provider,
		setProvider,
	};
};

export const ExchangeProvider = ({ children }: Properties) => {
	const context = useExchange();

	return <ExchangeContext.Provider value={context}>{children}</ExchangeContext.Provider>;
};

export const useExchangeContext = (): ReturnType<typeof useExchange> => {
	const context = React.useContext(ExchangeContext);
	if (context === undefined) {
		throw new Error("[useExchangeContext] Component not wrapped within a Provider");
	}
	return context;
};
