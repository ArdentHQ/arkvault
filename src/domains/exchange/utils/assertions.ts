import { AssertionError } from "assert";
import { Contracts } from "@ardenthq/sdk-profiles";

import { CurrencyData } from "@/domains/exchange/exchange.contracts";
import { ExchangeService } from "@/domains/exchange/services/exchange.service";

export function assertExchangeService(exchangeService?: ExchangeService): asserts exchangeService is ExchangeService {
	if (!(exchangeService instanceof ExchangeService)) {
		throw new AssertionError({
			message: `Expected 'exchangeService' to be ExchangeService, but received ${exchangeService}`,
		});
	}
}

export function assertExchangeTransaction(
	exchangeTransaction?: Contracts.IExchangeTransaction,
): asserts exchangeTransaction is Contracts.IExchangeTransaction {
	if (typeof exchangeTransaction?.orderId !== "function") {
		throw new AssertionError({
			message: `Expected 'exchangeTransaction' to be Contracts.IExchangeTransaction, but received ${exchangeTransaction}`,
		});
	}
}

export function assertCurrency(currencyData?: CurrencyData): asserts currencyData is CurrencyData {
	if (!currencyData || typeof currencyData !== "object" || !("coin" in currencyData && "name" in currencyData)) {
		throw new AssertionError({
			message: `Expected 'currencyData' to be CurrencyData, but received ${currencyData}`,
		});
	}
}
