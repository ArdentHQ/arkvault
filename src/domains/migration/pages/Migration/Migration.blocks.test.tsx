import React from "react";
import userEvent from "@testing-library/user-event";
import {
	MigrationHeader,
	MigrationHeaderExtra,
	MigrationNewMigrationMobileButton,
	ContractPausedAlert,
} from "./Migration.blocks";
import { render, screen } from "@/utils/testing-library";
import * as contexts from "@/app/contexts";

describe("MigrationHeader", () => {
	it("should render", () => {
		const onNewMigrationHandler = vi.fn();

		const { asFragment } = render(<MigrationHeader onNewMigration={onNewMigrationHandler} />);

		userEvent.click(screen.getByTestId("Migrations__add-migration-btn"));

		expect(onNewMigrationHandler).toHaveBeenCalled();

		expect(asFragment()).toMatchSnapshot();
	});
});

describe("MigrationHeaderExtra", () => {
	it("should render", () => {
		const onNewMigrationHandler = vi.fn();

		const { asFragment } = render(<MigrationHeaderExtra onNewMigration={onNewMigrationHandler} />);

		userEvent.click(screen.getByTestId("Migrations__add-migration-btn"));

		expect(onNewMigrationHandler).toHaveBeenCalled();

		expect(asFragment()).toMatchSnapshot();
	});
});

describe("MigrationNewMigrationMobileButton", () => {
	it("should render", () => {
		const onNewMigrationHandler = vi.fn();

		const { asFragment } = render(<MigrationNewMigrationMobileButton onNewMigration={onNewMigrationHandler} />);

		userEvent.click(screen.getByTestId("Migrations__add-migration-btn-mobile"));

		expect(onNewMigrationHandler).toHaveBeenCalled();

		expect(asFragment()).toMatchSnapshot();
	});
});

describe("ContractPausedAlert", () => {
	it("should render if contract is paused", () => {
		const useMigrationsSpy = vi.spyOn(contexts, "useMigrations").mockReturnValue({
			contractIsPaused: true,
		});

		render(<ContractPausedAlert />);

		expect(screen.getByTestId("ContractPausedAlert")).toBeInTheDocument();

		useMigrationsSpy.mockRestore();
	});

	it("should render empty if contract is not paused", () => {
		const useMigrationsSpy = vi.spyOn(contexts, "useMigrations").mockReturnValue({
			contractIsPaused: false,
		});

		render(<ContractPausedAlert />);

		expect(screen.queryByTestId("ContractPausedAlert")).not.toBeInTheDocument();

		useMigrationsSpy.mockRestore();
	});
});
