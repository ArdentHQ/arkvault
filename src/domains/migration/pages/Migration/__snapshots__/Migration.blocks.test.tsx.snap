// Vitest Snapshot v1

exports[`MigrationEmpty > should render 1`] = `
<DocumentFragment>
  <div
    class="mt-4 css-ycib48"
  >
    <div
      class="mx-auto px-8 md:px-10 lg:container"
    >
      <div
        class="rounded-lg border-solid border-theme-secondary-300 text-center leading-5 text-theme-secondary-text dark:border-theme-secondary-800 sm:border sm:p-6"
        data-testid="Migrations--empty-results"
      >
        <div>
          You don't have any migration transactions yet.
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MigrationHeader > should render 1`] = `
<DocumentFragment>
  <div
    class="border-theme-secondary-300 dark:border-theme-secondary-800 hasBorder css-835596"
  >
    <div
      class="mx-auto px-8 md:px-10 lg:container"
    >
      <div
        class="flex items-end justify-between bg-theme-background"
      >
        <div
          class="space-y-4"
        >
          <h1
            class="mb-0 text-2xl"
            data-testid="header__title"
          >
            Migrate to Polygon
          </h1>
          <div
            class="flex items-center text-theme-secondary-text"
            data-testid="header__subtitle"
          >
            ARK is moving to Polygon. Click "New Migration" to get started.
          </div>
        </div>
        <div>
          <div
            class="hidden py-4 md:py-0 sm:block sm:h-auto"
          >
            <button
              class="ml-6 px-5 md:py-3 py-2 css-1ac6umz"
              data-testid="Migrations__add-migration-btn"
              type="button"
            >
              <div
                class="flex items-center space-x-2"
              >
                New Migration
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MigrationHeaderExtra > should render 1`] = `
<DocumentFragment>
  <div
    class="hidden py-4 md:py-0 sm:block sm:h-auto"
  >
    <button
      class="ml-6 px-5 md:py-3 py-2 css-1ac6umz"
      data-testid="Migrations__add-migration-btn"
      type="button"
    >
      <div
        class="flex items-center space-x-2"
      >
        New Migration
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`MigrationNewMigrationMobileButton > should render 1`] = `
<DocumentFragment>
  <div
    class="sm:hidden"
  >
    <div
      class="css-2wyvkd"
    >
      <button
        class="css-1bx0hyc"
        data-testid="Migrations__add-migration-btn-mobile"
        type="button"
      >
        <div
          class="flex items-center space-x-2"
        >
          New Migration
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
